### PR TITLE
stgit: 1.5 -> 2.3.0

### PR DESCRIPTION
**Pull request by skyfmmf**
###### Description of changes

Version 2 of stgit was rewritten in Rust but in general offers the same feature set. We therefore do not require a Python build environment anymore but a Rust one. However, we overwrite the default build, check and install phases to use the Makefiles provided by stgit. In addition, the following changes are necessary:

- stgit now requires libcurl
- perl is required during build (to build docs correctly) and during check (to run the extensive test suite)

Changelog for v2.0.0: https://github.com/stacked-git/stgit/releases/tag/v2.0.0
Changelog for v2.3.0: https://github.com/stacked-git/stgit/releases/tag/v2.3.0
List of changelogs v2.0.0...v2.3.0: https://stacked-git.github.io/changelog/

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable: (checkPhase runs the extensive test suite and I tested a few commands locally)
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage) (nothing depends on stgit)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by jshholland — APPROVED** [review]
LGTM; tested on x86_64-linux and aarch64-linux.

**Comment by skyfmmf** [comment]
The PR needed a rebase as d994dabacef ("treewide: remove git-and-tools directory") was applied in the meantime, which moved the stgit package. The contents of the PR are still the same.

**Review by jshholland — APPROVED** [review]

File: pkgs/applications/version-management/stgit/default.nix [hunk 1]
```diff
@@ -1,32 +1,37 @@
...
     rev = "v${version}";
-    sha256 = "sha256-TsJr2Riygz/DZrn6UZMPvq1tTfvl3dFEZZNq2wVj1Nw=";
+    sha256 = "sha256-ab85aX1eYuyUYR4x3Bh0pRd5xUrxrrX/9ilnBSFv/SA=";
   };
+  cargoHash = "sha256-3RuexRvsekzcACTMqZw1701eNwenKgBFHZ4Apd84Lag=";
 
-  nativeBuildInputs = [ installShellFiles asciidoc xmlto docbook_xsl docbook_xml_dtd_45 python3Packages.setuptools ];
+  nativeBuildInputs = [ pkg-config installShellFiles asciidoc xmlto docbook_xsl docbook_xml_dtd_45 perl ];
+  buildInputs = [ libgit2 curl bzip2 ];
+  propagatedBuildInputs = [ git ];
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/stgit/default.nix:32]: Please use makeWrapper instead.

File: pkgs/applications/version-management/stgit/default.nix [hunk 2]
```diff
@@ -40,24 +45,20 @@ python3Packages.buildPythonApplication rec {
...
     "XMLTO_EXTRA=--skip-validation"
+    "PERL_PATH=${perl}/bin/perl"
   ];
 
-  buildFlags = [ "all" "doc" ];
+  buildPhase = [ "make" "all" ] ++ makeFlags;
 
-  checkTarget = "test";
-  checkFlags = [ "PERL_PATH=${perl}/bin/perl" ];
+  checkPhase = [ "make" "test" ] ++ makeFlags;
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/stgit/default.nix:59]: This should be checkTarget

File: pkgs/applications/version-management/stgit/default.nix [hunk 3]
```diff
@@ -40,24 +45,20 @@ python3Packages.buildPythonApplication rec {
...
 
-  installTargets = [ "install" "install-doc" "install-html" ];
+  installPhase = [ "make" "install" "install-man" "install-html" ] ++ makeFlags;
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/stgit/default.nix:61]: This should be kept installTargets. installPhase and checkPhase are multiline strings, not lists.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/applications/version-management/stgit/default.nix (hunk 1 above)
Comment (skyfmmf) [reply pkgs/applications/version-management/stgit/default.nix:32]: You mean, I should keep git in `buildInputs` and run makeWrapper/wrapProgram with `--prefix PATH : ${lib.makeBinPath [ git ]}` in `installPhase` instead?

File: pkgs/applications/version-management/stgit/default.nix (hunk 2 above)
Comment (skyfmmf) [reply pkgs/applications/version-management/stgit/default.nix:59]: I'm not sure how that would work. This now uses `rustPlatform.buildRustPackage` which ignores checkTargets. Instead, it just runs the Rust test then, which are very different than the integration tests behind `make test`.

For a Rust based program that use a Makefiles, the nixpkgs docs write the following: 

> Some packages may use custom scripts for building/installing, e.g. with a Makefile. In these cases, it’s recommended to override the buildPhase/installPhase/checkPhase. 

(from https://nixos.org/manual/nixpkgs/stable/#compiling-rust-applications-with-cargo, 17.31.1.6. Custom build/install-procedures)

Using checkTarget that automatically uses the makeFlags sounds cleaner, I'm just now sure how I would use that with Rust.

File: pkgs/applications/version-management/stgit/default.nix (hunk 3 above)
Comment (skyfmmf) [reply pkgs/applications/version-management/stgit/default.nix:61]: See my comment for checkPhase/checkTargets

**Comment by AndersonTorres** [comment]
Still needs rebase.

**Review by Janik-Haag — COMMENTED** [review]
@skyfmmf can you please rebase so we can get this merged?

**Comment by skyfmmf** [comment]
@Janik-Haag Sorry for loosing track of this for a while. 

I rebased, updated directly to the most recently released version and addressed the review comments.

File: pkgs/applications/version-management/stgit/default.nix (hunk 1 above)
Comment (skyfmmf) [reply pkgs/applications/version-management/stgit/default.nix]: Updated to use makeProgram

File: pkgs/applications/version-management/stgit/default.nix (hunk 2 above)
Comment (skyfmmf) [reply pkgs/applications/version-management/stgit/default.nix]: As this is the recommended way to package Rust programs with custom build processes, I'll stick with checkPhase. But I updated it to be a multiline string as you noted.

File: pkgs/applications/version-management/stgit/default.nix (hunk 3 above)
Comment (skyfmmf) [reply pkgs/applications/version-management/stgit/default.nix]: As this is the recommended way to package Rust programs with custom build processes, I'll stick with installPhase. But I updated it to be a multiline string as you noted.

**Comment by Janik-Haag** [comment]
Result of `nixpkgs-review pr 205220` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>stgit</li>
  </ul>
</details>

**Review by jshholland — APPROVED** [review]

**Review by Janik-Haag — APPROVED** [review]

**Comment by Janik-Haag** [comment]
Darwin tests are failing, if you know how to fix them fix them, if not ping the darwin-maintainers team or make darwin as broken since it's supposed to be working.

**Comment by skyfmmf** [comment]
@Janik-Haag I fixed the darwin tests. Apparently they need sysctl (to get the number of cores) and iconv to pass.

File: pkgs/applications/version-management/stgit/default.nix [hunk 4]
```diff
@@ -1,32 +1,48 @@
...
+    pkg-config installShellFiles makeWrapper asciidoc xmlto docbook_xsl
+    docbook_xml_dtd_45 perl
+  ];
+  buildInputs = [ curl ];
 
-  format = "other";
+  nativeCheckInputs = [
+    git perl
+  ] ++ lib.optionals stdenv.isDarwin [
+    darwin.system_cmds darwin.libiconv
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/stgit/default.nix:39]: this should be just libiconv

File: pkgs/applications/version-management/stgit/default.nix [hunk 5]
```diff
@@ -40,25 +56,28 @@ python3Packages.buildPythonApplication rec {
...
   ];
 
-  buildFlags = [ "all" "doc" ];
+  buildPhase = ''
+    make all ${makeFlags}
+  '';
+
+  checkPhase = ''
+    make test ${makeFlags}
+  '';
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/stgit/default.nix:74]: We don't need to manually define this, it should just work with makeFlags. Also runHook is missing

Comment (skyfmmf) [reply pkgs/applications/version-management/stgit/default.nix:74]: I just tried, it doesn't. But it turns out, most of the flags are not necessary anymore, I'll clean them up and add the runHooks with the next release of stgit

**Pull request merged**